c10/core/AllocatorConfig.h

```diff
@@ -0,0 +1,217 @@
+#pragma once
+
+#include <c10/core/DeviceType.h>
+#include <c10/util/Exception.h>
+#include <c10/util/llvmMathExtras.h>
+
+#include <array>
+#include <atomic>
+#include <mutex>
+#include <string>
+#include <vector>
+
+namespace c10::CachingAllocator {
+
+// "large" allocations may be packed in 20 MiB blocks
+const size_t kLargeBuffer = 20971520;
+
+/**
+ * Note [AllocatorConfig design]
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * This class configures the allocator for both device and host memory
+ * management. A single AllocatorConfig is used across all devices, such as CUDA
+ * and XPU, assuming that environment variables apply universally.
+ *
+ * It is designed to *ONLY* contain configuration options that can be set via
+ * environment variables.
+ *
+ * Naming Convention:
+ * - Public API names in AllocatorConfig should be device-generic. For example,
+ *     `use_release_lock_on_cudamalloc` is not ideal; instead, use
+ *     `use_release_lock_on_device_malloc` to maintain neutrality.
+ * - Members prefixed with `pinned_` are specific to the host/pinned allocator.
+ * - Environment variable names should also be device-generic to ensure
+ *     consistency across different hardware backends.
+ *
+ * Environment Variables:
+ * - The default environment variable for configuration is `PYTORCH_ALLOC_CONF`.
+ * - For backward compatibility, `PYTORCH_CUDA_ALLOC_CONF` is also supported
+ *     with lower priority.
+ */
+
+class C10_API AllocatorConfig {
+ public:
+  static AllocatorConfig& instance();
+
+  C10_DISABLE_COPY_AND_ASSIGN(AllocatorConfig);
+  AllocatorConfig(AllocatorConfig&&) = delete;
+  AllocatorConfig& operator=(AllocatorConfig&&) = delete;
+  ~AllocatorConfig() = default;
+
+  /* Device allocator settings */
```

Comment:
I feel like we want to add documentation for all of these?
Also you're confident all of these are relevant for all accelerators? A lot of these feel very cuda specific and I don't really see a benefit with moving them here?